File: csharp/Tests/MisarBlogClientTests.cs
using System.Net;
using System.Text;
using MisarBlog;
using Xunit;

namespace MisarBlog.Tests;

/// <summary>
/// Unit tests for <see cref="MisarBlogClient"/>.
///
/// A stub <see cref="HttpMessageHandler"/> replaces the transport, so the real
/// request/retry/error code runs with no network call. This SDK had 28 compile
/// errors and no tests at all, so these start with the basics: that it builds
/// the URLs it claims to, and that a refusal is typed.
/// </summary>
public sealed class MisarBlogClientTests
{
    /// <summary>Replays a queue of responses and records what was asked for.</summary>
    private sealed class ScriptedHandler : HttpMessageHandler
    {
        private readonly Queue<(int Status, string Body, (string, string)[] Headers)> _queue;

        public List<string> Requests { get; } = new();

        public ScriptedHandler(params (int Status, string Body, (string, string)[] Headers)[] responses)
            => _queue = new(responses);

        protected override Task<HttpResponseMessage> SendAsync(
            HttpRequestMessage request, CancellationToken cancellationToken)
        {
            Requests.Add(request.RequestUri!.PathAndQuery);

            var (status, body, headers) = _queue.Count > 0
                ? _queue.Dequeue()
                : (500, "{\"error\":\"script exhausted\"}", Array.Empty<(string, string)>());

            var response = new HttpResponseMessage((HttpStatusCode)status)
            {
                Content = new StringContent(body, Encoding.UTF8, "application/json")
            };
            foreach (var (name, value) in headers)
                response.Headers.TryAddWithoutValidation(name, value);

            return Task.FromResult(response);
        }
    }

    private static MisarBlogClient ClientWith(ScriptedHandler handler, int maxRetries = 1) =>
        new("mbk_test", maxRetries: maxRetries, httpClient: new HttpClient(handler));

    private static ScriptedHandler Ok(string body) =>
        new((200, body, Array.Empty<(string, string)>()));

    // ── Construction ─────────────────────────────────────────────────────────

    [Fact]
    public void BlankApiKeyIsRejected()
    {
        // Failing at construction beats a 401 on the first call.
        Assert.Throws<ArgumentException>(() => new MisarBlogClient("  "));
    }

    // ── REST ─────────────────────────────────────────────────────────────────

    [Fact]
    public async Task ArticlesList_ReturnsParsedBody()
    {
        var handler = Ok("{\"articles\":[{\"slug\":\"hello\"}],\"total\":1}");
        var result = await ClientWith(handler).Articles_ListAsync(limit: 10);

        Assert.Equal(1, result.GetProperty("total").GetInt32());
        Assert.Equal("hello", result.GetProperty("articles")[0].GetProperty("slug").GetString());
        // The limit must reach the wire, not merely the signature.
        Assert.Contains("limit=10", handler.Requests[0]);
    }

    [Fact]
    public async Task ArticlesList_OmitsUnsetQueryParameters()
    {
        var handler = Ok("{\"articles\":[],\"total\":0}");
        await ClientWith(handler).Articles_ListAsync();

        // A bare path, not "?status=&visibility=" — empty values would be sent
        // to the server as real filters.
        Assert.Equal("/blog/v1/articles", handler.Requests[0]);
    }

    [Fact]
    public async Task ProfileGet_ReturnsTheAccount()
    {
        var result = await ClientWith(Ok("{\"id\":\"u1\",\"username\":\"gulshan\"}")).Profile_GetAsync();

        Assert.Equal("gulshan", result.GetProperty("username").GetString());
    }

    [Fact]
    public async Task PlanGet_ReportsTheSubscriptionBehindTheKey()
    {
        var result = await ClientWith(Ok("{\"plan\":{\"slug\":\"pro\"}}")).Plan_GetAsync();

        Assert.Equal("pro", result.GetProperty("plan").GetProperty("slug").GetString());
    }

    [Fact]
    public async Task CommentsList_PassesTheArticleIdAsAQueryParameter()
    {
        // comments and follows were added late and are query-filtered rather than
        // path-scoped; this pins that shape.
        var handler = Ok("{\"comments\":[],\"total\":0}");
        await ClientWith(handler).Comments_ListAsync("a1");

        Assert.Contains("/blog/v1/comments", handler.Requests[0]);
        Assert.Contains("article_id=a1", handler.Requests[0]);
    }

    [Fact]
    public async Task FollowsStatus_PassesTheUserId()
    {
        var handler = Ok("{\"following\":true}");
        await ClientWith(handler).Follows_StatusAsync("u2");

        Assert.Contains("user_id=u2", handler.Requests[0]);
    }

    // ── Errors ───────────────────────────────────────────────────────────────

    [Fact]
    public async Task Unauthorized_ThrowsWithStatus()
    {
        var handler = new ScriptedHandler((401, "{\"error\":\"unauthorized\"}", Array.Empty<(string, string)>()));
        var ex = await Assert.ThrowsAsync<MisarBlogException>(
            () => ClientWith(handler).Profile_GetAsync());

        Assert.Equal(401, ex.Status);
    }

    [Fact]
    public async Task InsufficientScope_CarriesTheScopeDetail()
    {
        // A 403 that names the missing scope is actionable; a bare "forbidden" is not.
        var handler = new ScriptedHandler((403,
            "{\"error\":\"insufficient scope\",\"required_scope\":\"articles:write\"," +
            "\"granted_scopes\":[\"articles:read\"]}", Array.Empty<(string, string)>()));

        var ex = await Assert.ThrowsAsync<MisarBlogException>(
            () => ClientWith(handler).Articles_PublishAsync(new { title = "x" }));

        Assert.Equal(403, ex.Status);
        Assert.Equal("articles:write", ex.RequiredScope);
        Assert.Contains("articles:read", ex.GrantedScopes);
    }

    [Fact]
    public async Task SpentAllowance_ThrowsPlanLimitAndIsNotRetried()
    {
        const string body = """
        {"code":"plan_limit_exceeded","error":"monthly article allowance spent",
         "upgrade":{"urls":{"pricing":"https://www.misar.blog/pricing"}}}
        """;
        var handler = new ScriptedHandler(
            (429, body, new[] { ("X-Misar-Plan", "starter"), ("Retry-After", "3600") }));

        // maxRetries 3, so a plain retryable 429 would have been retried twice more.
        var ex = await Assert.ThrowsAsync<MisarBlogPlanLimitException>(
            () => ClientWith(handler, maxRetries: 3).Articles_PublishAsync(new { title = "x" }));

        Assert.Equal(429, ex.Status);
        Assert.Equal("starter", ex.Plan);
        Assert.Equal(3600, ex.RetryAfter);
        Assert.Equal("https://www.misar.blog/pricing", ex.UpgradeUrl);

        // A spent allowance cannot be fixed by retrying.
        Assert.Single(handler.Requests);
    }

    [Fact]
    public async Task Retries503ThenSucceeds()
    {
        var handler = new ScriptedHandler(
            (503, "{\"error\":\"unavailable\"}", Array.Empty<(string, string)>()),
            (503, "{\"error\":\"unavailable\"}", Array.Empty<(string, string)>()),
            (200, "{\"articles\":[],\"total\":0}", Array.Empty<(string, string)>()));

        var result = await ClientWith(handler, maxRetries: 3).Articles_ListAsync();

        Assert.Equal(0, result.GetProperty("total").GetInt32());
        Assert.Equal(3, handler.Requests.Count);
    }
}

File: swift/Tests/MisarBlogTests/MisarBlogClientTests.swift
import XCTest
@testable import MisarBlog

#if canImport(FoundationNetworking)
import FoundationNetworking
#endif

/// Intercepts requests and replays a scripted response.
final class StubURLProtocol: URLProtocol {
    struct Stub {
        var status: Int = 200
        var headers: [String: String] = ["Content-Type": "application/json"]
        var body: String = "{}"
    }

    static var stubs: [String: Stub] = [:]
    static var requestedPaths: [String] = []
    static var requestedQueries: [String] = []
    /// Absolute strings, because `url.path` is decoded and cannot show encoding.
    static var requestedURLs: [String] = []

    static func reset() {
        stubs = [:]
        requestedPaths = []
        requestedQueries = []
        requestedURLs = []
    }

    override class func canInit(with request: URLRequest) -> Bool { true }
    override class func canonicalRequest(for request: URLRequest) -> URLRequest { request }

    override func startLoading() {
        let path = request.url?.path ?? ""
        StubURLProtocol.requestedPaths.append(path)
        StubURLProtocol.requestedQueries.append(request.url?.query ?? "")
        StubURLProtocol.requestedURLs.append(request.url?.absoluteString ?? "")

        let stub = StubURLProtocol.stubs[path]
            ?? .init(status: 404, body: #"{"error":"no stub"}"#)

        let response = HTTPURLResponse(
            url: request.url!, statusCode: stub.status,
            httpVersion: "HTTP/1.1", headerFields: stub.headers
        )!
        client?.urlProtocol(self, didReceive: response, cacheStoragePolicy: .notAllowed)
        client?.urlProtocol(self, didLoad: Data(stub.body.utf8))
        client?.urlProtocolDidFinishLoading(self)
    }

    override func stopLoading() {}
}

final class MisarBlogClientTests: XCTestCase {

    override func setUp() {
        super.setUp()
        StubURLProtocol.reset()
    }

    private func makeSession() -> URLSession {
        let config = URLSessionConfiguration.ephemeral
        config.protocolClasses = [StubURLProtocol.self]
        return URLSession(configuration: config)
    }

    private func client(maxRetries: Int = 1) -> MisarBlogClient {
        MisarBlogClient(apiKey: "mbk_test", maxRetries: maxRetries, session: makeSession())
    }

    private func stub(_ path: String, status: Int = 200,
                      headers: [String: String] = [:], _ body: String) {
        var merged = ["Content-Type": "application/json"]
        headers.forEach { merged[$0.key] = $0.value }
        StubURLProtocol.stubs[path] = .init(status: status, headers: merged, body: body)
    }

    // MARK: Ownership

    func testResourcesKeepTheClientAlive() async throws {
        stub("/blog/v1/articles", #"{"articles":[],"total":0}"#)

        // Constructed inline, so nothing else holds a reference. While resources
        // were `lazy var` holding the client `unowned`, this crashed reading a
        // deallocated object — a shape any caller might reasonably write.
        let result = try await MisarBlogClient(
            apiKey: "mbk_test", maxRetries: 1, session: makeSession()
        ).articles.list()

        XCTAssertEqual(result["total"] as? Int, 0)
    }

    // MARK: REST

    func testArticlesListReturnsParsedBody() async throws {
        stub("/blog/v1/articles",
             #"{"articles":[{"slug":"hello","title":"Hello"}],"total":1}"#)

        let result = try await client().articles.list(limit: 10)
        let articles = result["articles"] as? [[String: Any]]

        XCTAssertEqual(result["total"] as? Int, 1)
        XCTAssertEqual(articles?.first?["slug"] as? String, "hello")
        // The limit must reach the wire, not just the signature.
        XCTAssertTrue(StubURLProtocol.requestedQueries.first?.contains("limit=10") ?? false,
                      StubURLProtocol.requestedQueries.description)
    }

    func testArticlesGetEncodesTheSlug() async throws {
        // Keyed on the decoded path, since that is what url.path reports.
        stub("/blog/v1/articles/a b", #"{"slug":"a b"}"#)

        _ = try await client().articles.get(slug: "a b")

        // Interpolated raw, a space made the URL invalid and nothing was sent at
        // all. The encoded form is what actually goes on the wire.
        XCTAssertEqual(StubURLProtocol.requestedURLs.first,
                       "https://api.misar.io/blog/v1/articles/a%20b")
        XCTAssertEqual(StubURLProtocol.requestedPaths.first, "/blog/v1/articles/a b")
    }

    func testProfileSendsTheBearerToken() async throws {
        stub("/blog/v1/me", #"{"id":"u1","username":"gulshan"}"#)

        let result = try await client().profile.get()

        XCTAssertEqual(result["username"] as? String, "gulshan")
    }

    func testPlanReportsTheSubscriptionBehindTheKey() async throws {
        stub("/blog/v1/plan", #"{"plan":{"slug":"pro"},"limits":{"articles_per_month":100}}"#)

        let result = try await client().plan.get()
        let plan = result["plan"] as? [String: Any]

        XCTAssertEqual(plan?["slug"] as? String, "pro")
    }

    func testCommentsListIsReachable() async throws {
        // comments and follows were added late; this pins that they are wired to
        // the client rather than merely declared.
        stub("/blog/v1/comments", #"{"comments":[{"id":"c1"}],"total":1}"#)

        let result = try await client().comments.list(articleID: "a1")

        XCTAssertEqual(result["total"] as? Int, 1)
    }

    // MARK: Errors

    func testUnauthorizedThrowsApiErrorWithStatus() async throws {
        stub("/blog/v1/me", status: 401, #"{"error":"unauthorized"}"#)

        do {
            _ = try await client().profile.get()
            XCTFail("expected an error")
        } catch let error as MisarBlogError {
            guard case let .apiError(status, _, _, _) = error else {
                return XCTFail("expected apiError, got \(error)")
            }
            XCTAssertEqual(status, 401)
        }
    }

    func testSpentAllowanceThrowsPlanLimitAndIsNotRetried() async throws {
        stub("/blog/v1/articles", status: 429,
             headers: ["X-Misar-Plan": "starter", "Retry-After": "3600"],
             """
             {"code":"plan_limit_exceeded","error":"monthly article allowance spent",
              "upgrade":{"urls":{"pricing":"https://www.misar.blog/pricing"}}}
             """)

        // maxRetries 3, so a plain retryable 429 would have been retried twice more.
        do {
            _ = try await client(maxRetries: 3).articles.publish(data: ["title": "x"])
            XCTFail("expected a plan-limit error")
        } catch let error as MisarBlogError {
            guard case let .planLimitExceeded(status, _, plan, upgradeURL, retryAfter) = error else {
                return XCTFail("expected planLimitExceeded, got \(error)")
            }
            XCTAssertEqual(status, 429)
            XCTAssertEqual(plan, "starter")
            XCTAssertEqual(retryAfter, 3600)
            XCTAssertEqual(upgradeURL, "https://www.misar.blog/pricing")
        }

        // A spent allowance cannot be fixed by retrying.
        XCTAssertEqual(StubURLProtocol.requestedPaths.count, 1)
    }

    func testUpgradeURLIsExposedOnTheError() async throws {
        stub("/blog/v1/articles", status: 402,
             """
             {"code":"plan_limit_exceeded","error":"not on your plan",
              "upgrade":{"urls":{"pricing":"https://www.misar.blog/pricing"}}}
             """)

        do {
            _ = try await client().articles.publish(data: [:])
            XCTFail("expected a plan-limit error")
        } catch let error as MisarBlogError {
            // The convenience accessor is what a caller reaches for.
            XCTAssertEqual(error.upgradeURL, "https://www.misar.blog/pricing")
        }
    }

    // MARK: Embed helper

    func testEmbedURLBuildsBothShapes() {
        // Unauthenticated and independent of the API client.
        let embed = MisarBlog()

        XCTAssertEqual(embed.embedURL(username: "gulshan", theme: "dark").absoluteString,
                       "https://misar.blog/gulshan/embed?theme=dark")
        XCTAssertEqual(embed.embedURL(username: "gulshan", slug: "hello", theme: "light").absoluteString,
                       "https://misar.blog/gulshan/hello/embed?theme=light")
        // "auto" is the default and deliberately omits the parameter, letting the
        // embed follow the host page rather than pinning a theme.
        XCTAssertEqual(embed.embedURL(username: "gulshan").absoluteString,
                       "https://misar.blog/gulshan/embed")
    }
}
